records/README.md

```diff
@@ -0,0 +1,220 @@
+
+### The Interface
+
+The IPRS interface is below. It has a few types and functions. We use the Go language to express it, but this is language agnostic.
+
+```go
+// Record is the base type. user can define other types that
+// extend Record.
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the rest of record
+  Value     Data // an opaque value
+}
+
+// Validator is a function that returns whether a record is valid.
+// Users provide their own Validator implementations.
+type Validator func(r *Record) (bool, error)
+
+// Marshal/Unmarshal specifies a way to code the record
+type Marshal(r *Record) ([]byte, error)
+type Unmarshal(r *Record, []byte) (error)
+```
+
+### Interface Example
+
+For example, Alice and Bob earlier could use the following interface:
+
+```protobuf
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the record
+  Expires   Data // datetime at which record expires
+  Value     Data // an opaque value
+}
+
+
+func Validator(r *Record) (bool, error) {
+  authorKey := recordSigningKey(r)
+
+  // always check the signature first
+  sigok := authorKey.Verify(r.Signature, signablePart(r))
+  if !sigok {
+    return false, errors.New("invalid signature. forged record?")
+  }
+
+  // check the expiration.
+  if r.Expires < time.Now() {
+    return false, errors.New("record expired.")
+  }
+
+  return true, nil
+}
+
+func Marshal(r *Record) ([]byte, error) {
+  return recordProtobuf.Marshal(r)
+}
+
+func Unmarshal(r *Record, d []byte) (error) {
+  return recordProtobuf.Unmarshal(r, d)
+}
+```
+
+## Example Record Types
+
+For ease of use, IPRS implementations should include a set of common record types:
+
+- signed, valid within a datetime range
+- signed, expiring after a Time-To-Live
+- signed, based on ancestry (chain)
+- signed, with cryptographic freshness
+
+
+### Signed, valid within a datetime range
+
+This record type uses digital signatures (and thus a PKI) and timestamps (and thus a TI). It establishes that a record is valid during a particular datetime range. 0 (beginning of time), and infinity (end of time) can express unbounded validity.
+
+### Signed, expiring after a Time-To-Live
+
+This record type uses digital signatures (and thus a PKI) and TTLs (and thus a TI). It establishes that a record is valid for a certain amount of time after a particular event. For example, an event may be "upon receipt" to specify that a record is valid for a given amount of time after a processor first receives it. This is equivalent to the way DNS sets expiries.
+
+### Signed, based on ancestry (chain)
+
+This record type uses digital signatures (and thus a PKI) and merkle-links to other, previous records. It establishes that the "most recent" record (merkle-ordered) is the most valid. This functions similar to a git commit chain.
```

Comment:
>  Or is the request going to have special information (e.g. “give me any records for that are descendents of ”, this seems to require too much local storage on the record-serving node).

It may be possible to bundle with bitswap replication / query notation -- which will be something like git ref notation and xpath... -- (things like `want <path>/**/*` or `want <path>~10` OR go the full xpath route... oof.)

>  This is probably going to be a case where we break the “should” recommendation for full record bundling, and just leave it up to the client to fetch any intervening records.

Yeah, agreed. already there's the case that a user will likely have the keys used to sign records with, so dont need to send. 

i wonder if it makes sense to come up with a "bundle" abstraction in bitswap (or other replication protocol) that expresses "I'm about to send you this graph (all the keys)", and then starts sending. meanwhile, the receiver -- after getting the first message -- can quickly ack the keys it does have and prevent the send. the simpler (but maybe less well-performing) option is to just wait till the user requests things.

#### just some examples to get people thinking:

![](https://www.evernote.com/l/AMZGfsahutRJa69sg4_VvHWPVM71nul_RPsB/image.png)

![](https://www.evernote.com/l/AMbM6MrByUFA3oablMjtiuJvGsulNFZBBScB/image.png)


@krl this is another concerns for _requiring_ records to _always be_ based on ancestry. the big one though, is that records can be lost, so the chain may not be complete and have to deal with that.